bob-apps/bin/bobc.rs

```diff
@@ -1,137 +1,421 @@
+        Regex::new(&st).unwrap()
+    }
+    fn get_filenames(self, keys: &KeyPattern) -> Box<dyn Iterator<Item = (u64, String)>> {
+        match keys {
+            KeyPattern::Single(val) => {
+                Box::new(iter::once((*val, self.inner.replace("{key}", &val.to_string()))))
+            }
+            KeyPattern::Range(r) => {
+                Box::new(r.clone().map(move |n| (n, self.inner.replace("{key}", &n.to_string()))))
+            }
+            KeyPattern::Multiple(v) => Box::new(
+                v.clone()
+                    .into_iter()
+                    .map(move |n| (n, self.inner.replace("{key}", &n.to_string()))),
+            ),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum KeyPattern {
+    Single(u64),
+    Range(RangeInclusive<u64>),
+    Multiple(Vec<u64>),
+}
+
+#[derive(Debug)]
+struct KeyName {
+    key: u64,
+    name: String,
+}
+
 #[tokio::main]
 async fn main() {
     env_logger::builder().filter_level(LevelFilter::Info).init();
     let matches = get_matches();
-    if let (sc, Some(sub_mathes)) = matches.subcommand() {
-        let key = get_key_value(sub_mathes);
-        match sc {
-            "put" => {
-                let value = sub_mathes.value_of("size").unwrap();
-                let size = value.parse().expect("size must be usize");
-                let addr = sub_mathes
-                    .value_of("uri")
-                    .expect("has default value")
-                    .parse()
-                    .expect("wrong format of url");
-                info!("PUT key: \"{:?}\" size: \"{}\" to \"{}\"", key, size, addr);
-                put(key, size, addr).await;
+    let app_args = AppArgs::from_matches(matches);
+
+    let addr = NetConfig::from_args(&app_args).get_uri();
+    let mut client = BobApiClient::connect(addr).await.unwrap();
+    match app_args.subcommand.as_str() {
+        PUT_SC => {
+            let keys_names = match (app_args.file_pattern.unwrap(), app_args.key_pattern) {
+                (FilePattern::WithRE(re, dir), None) => {
+                    prepare_put_from_pattern(&re.get_regex(), &dir).await
+                }
+                (FilePattern::WithRE(re, _), Some(key_pattern)) => {
+                    Box::new(re.get_filenames(&key_pattern).map(|(key, name)| KeyName {key,name}))
+                }
+                (FilePattern::WithoutRE(path), Some(KeyPattern::Single(key))) => {
+                    Box::new(iter::once(KeyName {
+                        key,
+                        name: path.to_string()
+                    }))
+                }
+                (FilePattern::WithoutRE(path), Some(KeyPattern::Range(r))) => {
+                    Box::new(r.map(move |key| KeyName {
+                        key,
+                        name: path.to_string()
+                    }))
+                }
+                (FilePattern::WithoutRE(path), Some(KeyPattern::Multiple(v))) => {
```

Comment:
Looks like `KeyPattern` struct should allow to iterate over its values. In this way you can merge last 3 branches into one

Reply:
`KeyPattern` can support `IntoIterator` trait (https://doc.rust-lang.org/std/iter/trait.IntoIterator.html)

Reply:
Done: [a37f10e](https://github.com/qoollo/bob/pull/562/commits/a37f10e3d7e8587a5169d396c1594248c26faeb0)